File: four/part2.rb
require 'set'

sum=0
row=0
multipliers=Hash.new(1)
while (line = gets)
  sum+=multipliers[row]
  winners, numbers = line.chomp.match(/: (.*?) \| (.*)/).captures
  winners = Set.new(winners.split.map(&:to_i))

  wins = numbers.split.map(&:to_i).count { |n| winners.include?(n) }
  (1..wins).each do |i|
    multipliers[row+i] = multipliers[row+i]+multipliers[row]
  end
  row+=1
end
puts sum

File: four/part2.py
import sys
sum=0
multipliers={}
row=0
for line in sys.stdin.readlines():
    sum+=multipliers.get(row,1)
    line = line[line.find(':') + 2:-1].replace('  ',' ')
    halves = line.split(' | ')
    print(halves)
    winners = set(halves[0].split(' '))
    wins=0
    for num in halves[1].split(' '):
        if num in winners:
            print(f'{num} is winner')
            print(f'multiplier:{multipliers.get(row,1)} inc mulitplier for row={row+wins}->{multipliers.get(row+wins,1)}')
            wins+=1
            multipliers[row+wins] = multipliers.get(row+wins,1)+multipliers.get(row,1)
    # sum += card_total
    row+=1

print(sum)
